### ISSUE TITLE
Mishandling of Empty Metals Version in Workspace

### ISSUE THREAD

**Issue by vzmerr** [scalameta/metals-vscode#1064]

**Comment by tgodzik** [comment]
I transferred it to metals-vscode, since this is a problem only with the VS Code extension.

**Comment by tanishiking** [comment]
When I set `0.11.8-SNAPSHOT` on `UserConfiguration` and leave it empty on `WorkspaceConfiguration` the `workspaceConfig.inspect("serverVersion")` gives me:
```
defaultLanguageValue: undefined
defaultValue: '0.11.7+19-1405f4de-SNAPSHOT'
globalLanguageValue: undefined
globalValue: '0.11.8-SNAPSHOT'
key: 'metals.serverVersion'
languageIds: undefined
workspaceFolderLanguageValue: undefined
workspaceFolderValue: undefined
workspaceLanguageValue: undefined
workspaceValue: ''
```

While `workspaceConfig.get("serverVersion")` returns `""`.  `workspaceValue` should be `undefined` instead of `''`.

**Comment by tanishiking** [comment]
We should be able to fix this by `config.inspect` and config override manually like https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration but it should be a problem in VSCode, maybe there's a issue or something in VSCode side 🤔

**Comment by tgodzik** [comment]
Yeah, but on the UI if you remove the setting it translates it to `metals.serverVersion = ""`

I think it should be possible to fix it using a schema.

**Comment by tanishiking** [comment]
I tested some

- Add `"minLength"` to schema: it doesn't work
  - we just can't write an empty string on UI
  - if we remove the workspace setting somehow, it will `metals.serverVersion = ""` instead of `undefined`.
- I realized that `metals.serverVersion = undefined` (in workspace configuration) iff `metals.serverVersion === ${defaultValue}` !
  - So, if we wanna set to default when it's empty, we have to set default as "" (or just remove `default` from the schema)

**Comment by tanishiking** [comment]
This is definitely VSCode's UI issue, but not sure what behavior is best for all of VSCode users 🤔

**Comment by vzmerr** [comment]
> In the workspace config below, the workspace config is undefined (!!!), so if we set something in User Configuration VSCode will honor the User Configuration.

Thank you so much for taking the time to investigate this @tanishiking.

Now I wonder if we can programmatically trigger `reset setting` if the field has empty string 🧐

**Comment by vzmerr** [comment]
For now, is it a good idea to add a short description to the setting saying that 

> make sure this setting is also defined in the Workspace tab.

?

**Comment by tanishiking** [comment]
> Now I wonder if we can programmatically trigger reset setting if the field has empty string 🧐
It is possible, but I think it's too hacky and we shouldn't do that.

> For now, is it a good idea to add a short description to the setting saying that
> make sure this setting is also defined in the Workspace tab.

Not sure it helps 🤔 Becuase the root cause of this is it's hard to notice it's default or not. Even there are some words like "make sure this setting is also defined... in Workspace tab" -> go to Workspace settings and see it's empty string -> but they can't tell it is comprehended as an empty string rather than "no setting".
Also, that kind of message may confuse users.

The root cause is the misconception that "empty string means no setting", which should be improved in VSCode UI, I believe.

I think the diff in https://github.com/scalameta/metals-vscode/issues/1064#issuecomment-1185151321 could mitigate the issue, what do you think? 

---

> Did you considered opening an issue on VSCode repository and explaining this problem? If there's reset button then it might be intentional behavior but it's better to ask them.

Yeah, I'm considering opening an issue in VSCode, I also think this is intended behavior (because if VSCode interprets the empty string as `undefined`, it also would be a problem) but it's worth improving somehow. I will do that :)

**Comment by tanishiking** [comment]
I unassigned because we haven't decided how to deal with this issue, discuss the solution and let me assign to someone if what we gonna do 👍

**Comment by vzmerr** [comment]
> in Workspace tab" -> go to Workspace settings and see it's empty string -> but they can't tell it is comprehended as an empty string rather than "no setting".

Well, I guess then they would most probably think it is undefined and try to copy paste the setting from the User tab, wouldn't they?

**Comment by tanishiking** [comment]
> Well, I guess then they would most probably think it is undefined and try to copy paste the setting from the User tab, wouldn't they?

I personally don't think so, because VSCode already says something when we have something (including an empty string on workspace) but we still misunderstand that. That's why I think these messages won't help.

<img width="550" alt="Screen Shot 2022-07-15 at 18 09 33" src="https://user-images.githubusercontent.com/9353584/179194161-aeee9bb4-1c05-42ab-a01d-ed486b1486e9.png">

**Comment by tgodzik** [comment]
I wonder if `minLength` wouldn't actually help at least a bit here. Since you would be able to save settings with an empty string and since it's usually the advanced users they would know that they need to remove it from json.

**Comment by dos65** [comment]
I've checked that `minLength` doesn't work.
Also, found an issue about that - https://github.com/microsoft/vscode/issues/151167 and they suggest just to check such setting on extension side using `inspect` :disappointed:
